### ISSUE TITLE
ProofGeneral incorrectly parses unterminated strings in comments in Coq

### ISSUE THREAD

**Issue by JasonGross** [ProofGeneral/PG#37]
```
$ cat src/foo.v
Check (* "a *) True.
$ coqc src/foo.v
Warning: Not interpreting "*)" as the end of current non-terminated comment because it occurs in a non-terminated string of the comment.
File "D:\Documents\GitHub\fiat\src/foo.v", line 1, characters 9-21:
Syntax Error: Lexer: Unterminated string
```
But PG thinks this is fine.

**Comment by cpitclaudel** [comment]
Is that documented behaviour for Coq? Otherwise, I'd file this as a Coq bug, regardless of the nice warning message.

**Comment by JasonGross** [comment]
I'm not sure if it's documented, but, in some sense, Coq can be considered "Right", because it preserves the invariant that if `TERM` is a valid program, then `(* TERM *)` is a comment.  (Consider trying to comment out `Definition end_comment := "*)"%string.`)

**Comment by cpitclaudel** [comment]
Doesn't the reverse work with preserving the invariant that `"TERM"` should be a string? Of course it doesn't work due to nesting, but oh well.

**Comment by JasonGross** [comment]
Well, stringifying terms is already broken (because nesting), but more work seems to have gone into the commenting invariant.  (And, anyway, it seems like "comment this block of code" is a more frequent operation than "stringify this block of code")

**Comment by cpitclaudel** [comment]
I don't know. "Comment this invalid block of code" is a much more common operation for me. Isn't it for you?

**Comment by Matafou** [comment]
This is inspired from ocaml, see below, so this is definitely a coq feature.

Let us consider this as non urgent bug. Changing the way comments and
string interact may be dangerous as pg code is full of subtle calls to
bla-syntactic-context.

P.
_________________________

~/tmp$ cat > foo.ml
let (* "a *) x = 3;;
~/tmp$ ocamlc foo.ml

File "foo.ml", line 1, characters 4-6:
Error: This comment contains an unterminated string literal
File "foo.ml", line 1, characters 7-8:
Error: String literal begins here




2016-01-20 17:15 GMT+01:00 Clément Pit--Claudel <notifications@github.com>:

**Comment by cpitclaudel** [comment]
You're right Pierre, and Xavier confirms the point is be able to comment out any valid piece of code, without single-line comments (`//`).

**Comment by Matafou** [comment]
For the record: we do see the warning in pg, but pg waits forever the
answer of the unfinished command.
This is an example of a script that PG cannot deal with whereas
coq(top,c,ide) can.

Is it really annoying Jason?

P.

2016-01-20 15:38 GMT+01:00 Jason Gross <notifications@github.com>:

**Comment by JasonGross** [comment]
The annoying part is that it's a bit confusing to debug a long file that PG thinks is perfectly fine, but coqc does not, especially having never seen the error message before.

**Comment by cpitclaudel** [comment]
Ah, so this is the opposite as what Pierre described, right? Works in PG but not coqc.

**Comment by JasonGross** [comment]
Yes, the "works in PG but not in coqc" is much harder to debug, and therefore more annoying.

**Comment by Matafou** [comment]
I don't understand. PG does not accept the script given by Jason. The
locked region stays in "unprocessed" color and I never get coq answer.

P.





2016-02-07 15:58 GMT+01:00 Jason Gross <notifications@github.com>:

**Comment by JasonGross** [comment]
Oh, ugh, maybe what happened is that I processed the file, later edited the comment incorrectly, and then got confused.  Is there anything sane that can be done to catch this failure mode (without reverting the buffer when you edit a comment)?

**Comment by Matafou** [comment]
menu PG/interrupt prover, and then going back to a sane line should do it.
P.

2016-02-08 22:16 GMT+01:00 Jason Gross <notifications@github.com>:

**Comment by JasonGross** [comment]
No, I want it to happen automatically; I shouldn't have to notice that something is wrong to get PG to complain about it (imagine if you had to tell Coq that there should be a type error somewhere before it would tell you about it).

**Comment by cpitclaudel** [comment]
Indeed; I think this is just a question of resource allocation; this particular bug is going to be a pain to fix, but of course we'd welcome a patch :)
I don't think I'll have time to work on this very soon. IIRC font-lock lets you add test properties before doing its syntactic fontification pass, to override too-simplistic syntax table hints. This would probably work here.

**Comment by Matafou** [comment]
There are pleny of ugly things that can be done when editing a comment in the locked zone. You can add a *) inside the comment for example(without a corresponding (* ).
Ideally we would highlight the dangerous patterns, but they are difficult to detect.
I am afraid we will just let users responsible for keeping valid comments when doing that...

**Comment by Matafou** [comment]
More on this "bug".

There are several issues:

1) Jason were right we miss ill-formed comment:
```
Definition bar:=3.
 (* "x *)
Definition foo:=3.
```
is OK for PG but not for coqc.

2) But even when we detect changes in comments the locked region is not updated.
Rty this:
```
Definition bar:=3.
 (* x *)
Definition foo:=3.
```
Now add a closing `*)` after x. The seconf `*)` is now font-locked as a non comment part, but the locked region stays unchanged.

Both bugs looks fixable. Anyone willing to have a look? I won't have time for that until months I am afraid.

**Comment by cpitclaudel** [comment]
I don't have much time for this either. How do you think the first one can be fixed? A local syntax table?

**Comment by Matafou** [comment]
I don't know, but maybe tuareg mode already does it? I ll check.

**Comment by Blaisorblade** [comment]
> I don't understand. PG does not accept the script given by Jason. The locked region stays in "unprocessed" color and I never get coq answer.

I've had a script I got to fully pass in ProofGeneral and wasn't parsed by Coq. There was some back-and-forth with some confusing behavior (including some hang), but I've never edited the unbalanced comment (which I only found later).

~~~
Beyond this, there's another mismatch between ProofGeneral and Coq: a script with unfinished proofs (without a `Qed.`, and followed by another statement) works in ProofGeneral and gives an error with `coqc`:

```
$ coqc -I . Foo.v
Error: There are pending proofs
```

This is also hard to debug—in my case, I did it by a slow bisection, and luckily it happened before my really slow automatic proofs, so this comment applies:

> Yes, the "works in PG but not in coqc" is much harder to debug, and therefore more annoying.

**Comment by Blaisorblade** [comment]
PS: should I report a separate issue about `Qed` or is it known? I can't find relevant issues in the tracker.

**Comment by JasonGross** [comment]
Report a separate issue for the Qed; the comment issue described here is a bug in PG; PG parses the document differently from coqtop and coqc, and this causes unexpected behavior.  Your issue is that PG doesn't tell you if your file is unfinished (has unterminated proofs or open sections), but coqtop acts the same way as PG here; it's not clear when or how PG should give you this information (and you can discover it at any time by seeing what happens if you do `Show Proof.` or `End foo.`; no need to bisect), so it's more properly a feature request.  I'd actually suggest making a feature request on the Coq bug tracker, asking for the error message to tell you what line the proof started on, and possibly the theorem name.

**Comment by Blaisorblade** [comment]
FWIW, there's a possible upstream feature request/bug here too: `coqc` could report where the unterminated string starts (in my case, the reported position was the document end), just like I'd expect from other languages.

The required reports are at:
https://github.com/ProofGeneral/PG/issues/154
https://coq.inria.fr/bugs/show_bug.cgi?id=5275
